### DISCUSSION TITLE
neovim + clangd  Unknown argument: '-fno-reorder-functions'

### DISCUSSION THREAD

**Discussion by cbrake** [zephyrproject-rtos/zephyr#57755] [Q&A]
I'm using Neovim with clangd language server -- it is generally working great. The build process generates `compile_commands.json` and `.clang-format` is provided -- very well done!

I am seeing one issue where `-fno-reorder-functions` is is passed as an argument to GCC. This causes the following error in  Neovim, generated by the clangd language server:

`Unknown argument: '-fno-reorder-functions'`

This is enabled in CMakeLists.txt:

```
if(NOT CMAKE_C_COMPILER_ID STREQUAL "Clang" AND
   NOT CMAKE_C_COMPILER_ID STREQUAL "IntelLLVM" AND
   NOT CMAKE_C_COMPILER_ID STREQUAL "ARMClang")
  # GCC assumed
  zephyr_cc_option(-fno-reorder-functions)

  # GCC 11 and above may generate a warning when dereferencing a constant
  # address pointer whose address is below the value specified by the
  # `min-pagesize` parameter (defaults to 0x1000). The `min-pagesize` parameter
  # is set to 0 such that GCC never generates any warnings for the constant
  # address pointers. For more details, refer to the GCC PR99578.
  zephyr_cc_option(--param=min-pagesize=0)

  if(NOT ${ZEPHYR_TOOLCHAIN_VARIANT} STREQUAL "xcc")
    zephyr_cc_option(-fno-defer-pop)
  endif()
else()
  # Clang produces false positive vla warnings
  zephyr_cc_option(-Wno-vla)
endif()
```

So we are using GCC to build and Clangd for the language server, and it appears not all arguments are compatible. What is the best solution to this?

**Comment by cbrake** [comment]
From https://github.com/clangd/clangd/issues/662

adding the following to a `.clangd` config file removed this error.

```
CompileFlags:
  Add: -Wno-unknown-warning-option
  Remove: [-m*, -f*]
```

Would it make sense to add this `.clangd` file to the zephyr repo?

**Reply by cbrake** [reply]
I see this is already ignored by in `.gitignore` so I guess it is up to users to find solutions like this -- perhaps because .clangd is needed for user-specific settings. Maybe it would make more sense to document things like this or provide a sample .clangd file with common fixes.

**Reply by kraj** [reply]
> From [clangd/clangd#662](https://github.com/clangd/clangd/issues/662)
> 
> adding the following to a `.clangd` config file removed this error.
> 
> ```
> CompileFlags:
>   Add: -Wno-unknown-warning-option
>   Remove: [-m*, -f*]
> ```
> 
> Would it make sense to add this `.clangd` file to the zephyr repo?
`-fno-reorder-functions` is gcc specific option which is not available in llvm/clang and thats what its complaining about.
I think it would be good to support clangd language server but unless thats supported adding a .clangd file maybe not a good thing to do as it wont be tested regularly, so maybe adding a CI step to test use of clangd and then appropriate .clangd file might be a long term maintainable solution.

**Comment by szczys** [comment]
Thanks for the clangd pointers! Since this is an issue for every Zephyr project I open, I solved this at the system-wide level:

1. Create directory in the Linux config home
    ```
    $ mkdir -p ~/.config/clangd
    ```
2. Add `~/.config/clangd/config.yml` file and populate it with the workaround for the error

    ```
    CompileFlags:
      Add: -Wno-unknown-warning-option
      Remove: [-m*, -f*]
    ```

**Reply by ho-ho-ho** [reply]
Just had the same problem and this post was very helpful, but there is one small mistake: clangd expects the file to be called `config.yaml`. Using `config.yml` did not work.
